File: leetcode/src/romannumeral.cpp
/*
 * write functions to convert integer2roman and roman2integer
 * I, II, III, IV, V, VI, VII, VIII, IX, X
 * note: in Roman numeral's definition, integer >= 3999 is invalid
 * */
#include "romannumeral.h"

/*
 * convert integer to Roman numeral string
 * */
string Solution::int2Roman(int num){
    char res[MaxSize];
    memset(res, 0, sizeof(MaxSize));

    int j = 0;
    int i = tags - 1;   // scan optional tag from greatest to least

    while(num > 0){
        if(num < digits[i]){
            --i;
        }else{
            res[j++] = alphas[i][0];

            if(i & 1){  // odd indexed Roman numeral tag has two chars
                res[j++] = alphas[i][1];
            }

            num -= digits[i];
        }
    }
    res[j] = '\0';

    return string(res);
}

/*
 * Roman numeral string places most significant numeral in most left
 * */
int Solution::roman2Int(const string& str){ // split string by fixed segment, plus to sum
    const int n = str.size();

    int res = 0;
    int i = 0;
    int j = tags - 1; // scan optional tag from greatest to least

    while(i < n && j >= 0){
        if(j & 1){ // to match two chars
            if(i < n-1 && str[i] == alphas[j][0] && str[i+1] == alphas[j][1]){
                res += digits[j];
                i += 2;
            }else{
                --j;
            }
        }else{  // to match one char
            if(str[i] == alphas[j][0]){
                res += digits[j];
                ++i;
            }else{
                --j;
            }
        }
    }

    return res;
}

const int Solution::digits[13] = {1, 4, 5, 9, 10, 40, 50, 90, 100, 400, 500, 900, 1000};

const char Solution::alphas[13][3] = {
        "I",   //1
        "IV",  //4
        "V",   //5
        "IX",  //9
        "X",   //10
        "XL",  //40
        "L",   //50
        "XC",  //90
        "C",   //100
        "CD",  //400
        "D",   //500
        "CM",  //900
        "M"    //1000
};

/* unit test is in ../cpp_unittest/romannumeral_unittest */

File: leetcode/java/romannumeral.java
/*
 * converion between Roman numerals and integers
 * I  - 1
 * IV - 4
 * V  - 5
 * IX - 9
 * X  - 10
 * XL - 40
 * L  - 50
 * XC - 90
 * C  - 100
 * CD - 400
 * D  - 500
 * CM - 900
 * M  - 1000
 *
 * note that: in Roman numeral's defition, integer >= 3999 is invalid
 * */
import java.io.*;
import java.util.*;

public class romannumeral{
    int[] digits = {1,4,5,9,10,40,50,90,100,400,500,900,1000};
    String[] alphas = {
        "I",
        "IV",
        "V",
        "IX",
        "X",
        "XL",
        "L",
        "XC",
        "C",
        "CD",
        "D",
        "CM",
        "M"
    };
    public String int2Roman(int num){
        StringBuilder builder = new StringBuilder();
        int j=0, i=12;
        while(num > 0){
            if(num < digits[i]){
                --i;
            }else{
                builder.append(alphas[i].charAt(0));
                if(i % 2 == 1){ // i%2 == 1
                    builder.append(alphas[i].charAt(1));
                }
                num -= digits[i];
            }
        }
        return builder.toString();
    }

    public int roman2Int(String s){
        int res=0, n=s.length(), i=0, j=12;
        while(i < n && j>=0){
            if(j % 2 ==1){  //to match 2 chars
                if(i<n-1 && s.substring(i,i+2).equals(alphas[j])){  //String.substring(i,j) [i,j)
                    res += digits[j];
                    i += 2;
                }else{
                    --j;
                }
            }else{          //to match 1 char
                if(s.substring(i,i+1).equals(alphas[j])){
                    res += digits[j];
                    ++i;
                }else{
                    --j;
                }
            }
        }
        return res;
    }

    public void test_01(){
        Scanner scan = new Scanner(System.in);
        while(true){
            System.out.println("please input integer to convert:");
            String str = scan.nextLine().trim();
            if(str.isEmpty())    break;
            int n = Integer.parseInt(str);
            System.out.println(int2Roman(n));
        }
    }

    public void test_02(){
        Scanner scan = new Scanner(System.in);
        while(true){
            System.out.println("please input Roman numerals in upper case to convert:");
            String str = scan.nextLine().trim();
            if(str.isEmpty())    break;
            System.out.println(roman2Int(str));
        }
    }

    public static void main(String[] args){
        romannumeral rn = new romannumeral();
        rn.test_02();
    }
}
